### DISCUSSION TITLE
keep-alive() on a python destination?

### DISCUSSION THREAD

**Discussion by petiepooo** [syslog-ng/syslog-ng#4098] [Q&A] [answered]
When I use an external program as my destination in syslog-ng OSE v3.37, I am able to set keep-alive(yes) so that it continues executing during a config reload (not a restart).  However, that doesn't seem to be an option when I have a python class as my destination.  A config reload seems to always reinitialize the class instance.

Is keeping python class instances alive during reloads a possibility?  I'm running a process that reads syslogs, parses information out of them, and turns them into commands for another process connected via stdio, and I need it to keep state during a config reload.  If nothing in the config's python block has changed, I would love to see syslog-ng keep it alive during a reload rather than having it deinit() and init() each time.

**Comment by bazsi** [comment, marked as answer]
Hi,

I think you could create your own private module and store globals there.
That module will not get wiped out, so any global state would be kept.

The driver instance is reinitialized as you suggest, but the same happens
with drivers written in C, so each of those which needs to keep state does
that explicitly.

Bazsi

On Fri, Aug 5, 2022 at 8:46 PM Pete ***@***.***> wrote:

> When I use an external program as my destination in syslog-ng OSE v3.37, I
> am able to set keep-alive(yes) so that it continues executing during a
> config reload (not a restart). However, that doesn't seem to be an option
> when I have a python class as my destination. A config reload seems to
> always reinitialize the class instance.
>
> Is keeping python class instances alive during reloads a possibility? I'm
> running a process that reads syslogs, parses information out of them, and
> turns them into commands for another process connected via stdio, and I
> need it to keep state during a config reload. If nothing in the config's
> python block has changed, I would love to see syslog-ng keep it alive
> during a reload rather than having it deinit() and init() each time.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/syslog-ng/syslog-ng/discussions/4098>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFOK5XWZYBY42JKU7RK5BDVXVORPANCNFSM55W6TKVQ>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 
Bazsi

**Reply by petiepooo** [reply]
I'm a little lost, as I have never developed or used private modules in syslog-ng.  Could that private module for storing globals be written in python and also defined within a python{...}; block in syslog-ng's config file?

**Reply by petiepooo** [reply]
I see the following in the docs:

> Caution:
If you reload syslog-ng OSE, existing Python objects are destroyed, therefore the context and state information of Python blocks is lost. Log rotation and updating the configuration of syslog-ng OSE typically involves a reload.

So I assume this is not as simple as defining another python{...}; block to hold the globals one wants to persist between reload.  Having to develop and load a separate module kind of defeats the biggest benefit to me of consolidating the python code for the destination into the syslog-ng configuration.  I might as well stick with the python script as an external program with keep-alive(yes) set.

**Comment by bazsi** [comment]
the binding should be a lot faster than an external script reparsing the
messages coming from its stdin.

you can also create a new module programatically:
https://stackoverflow.com/questions/25602885/porting-new-module-to-python3

On Sun, Aug 7, 2022 at 8:01 PM Pete ***@***.***> wrote:

> I see the following in the docs:
>
> Caution:
> If you reload syslog-ng OSE, existing Python objects are destroyed,
> therefore the context and state information of Python blocks is lost. Log
> rotation and updating the configuration of syslog-ng OSE typically involves
> a reload.
>
> So I assume this is not as simple as defining another python{...}; block
> to hold the globals one wants to persist between reload. Having to develop
> and load a separate module kind of defeats the biggest benefit to me of
> consolidating the python code for the destination into the syslog-ng
> configuration. I might as well stick with the python script as an external
> program with keep-alive(yes) set.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/syslog-ng/syslog-ng/discussions/4098#discussioncomment-3344323>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFOK5U433IYCT3N4NWHBCLVX72ZHANCNFSM55W6TKVQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
Bazsi

**Reply by petiepooo** [reply]
Ah, I think I understand what you're talking about now.  Thank you.
